format/IRFunction.fbs

```diff
@@ -0,0 +1,68 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+include "Schema.fbs";
+
+namespace org.apache.arrow.ir.flatbuf;
+
+// A unique identifier for a particular function definition. 
+table FunctionId {
+  // the function description identifier.
+  id: uint32;
+  
+  // The origin of the function definition. Should be mapped to a domain such as "org.apache.arrow"
+  // The github definition paths for definitions are defined in 
+  // github.com/apache/arrow/format/functions.txt. Org 0 == Apache Arrow canonical definitions.
+  // Organizations < 1B are canonical organizations. For private functions, use
+  // Org id > 1B
```

Comment:
Is this 1 Byte or 1 Billion? How are canonical organizations defined outside of `org.apache.arrow`?

Reply:
1 billion. I was thinking that we'd have a doc on the Arrow git that manages the public orgs with pointers to those orgs definitions. Such as:
|Id  |Organization   |Path|
|---|---|---|
|0|org.apache.arrow|https://github.com/apache/arrow/base_sql_functions.yaml|
|1|org.bar.com|https://github.com/bar/repo1/the_missing_arrow_functions.yaml|
|2|org.foo.com|https://github.com/foo/repo1/stats_functions.yaml|
|3|org.random.com|https://random.com/playful_functions.yaml|

Reply:
Is the idea to use something like this in the physical plan or logical plan or both?

Reply:
I would expect both. I can come up with scenarios where you might want them namespaced (logical vs physical) but it feels like premature extensibility.

Reply:
I'm concerned about having namespaced functions in the logical plan that aren't user defined functions.

For example, there's already functions and engines implemented in both C++ and Rust. If someone builds a logical plan from say a Python or R library, I imagine that logical plan should be able to be executed by either the C++ or Rust engine and that engine could choose which function implementation they should map to when building their physical plan.

I.E. say there's some arbitrary function that is standardized like `functionA` and there's both a C++ and Rust implementation of it. Would they have the same function IDs in different orgs or would they have the same org with different function IDs? How would an IR producer allow arbitrarily targeting either implementation?